### DISCUSSION TITLE
Add posibility to attach passport context to upgrade request

### DISCUSSION THREAD

**Discussion by Morishiri** [enisdenjo/graphql-ws#330] [Q&A] [answered]
### Story

As a server I want to be able to authenticate users and attach user context to request.

I want to be able to parse the incoming websocket request the same way as regular HTTP request and allow e.g. passport to attach user to request object so it can be later used in graphql `Subscription` context.

### Acceptance criteria

- Server is able to obtain user from the incoming request via passport middleware

### Additional context

In my case `passport` uses `oidc` strategy and cookies, I cannot simply attach the Bearer token to `connectionParams` as mentioned in the docs.

I modified the proposed `useServer` call to this form:
```
useServer(
    {
      schema,
      context: async (ctx) => {
        // I want to be able to get request.user here!
      },
      onConnect: async (ctx) =>
        new Promise((resolve) => {
          // Is stored passport.initialize() from above
          passportInitialize(ctx.extra.request, {}, () => {
            // Is stored passport.session() from above
            passportSession(ctx.extra.request, {}, () => {
              // request here has `_passport` property, but user is not there, why?
              resolve(true);
            });
          });
        }),
    },
    wsServer,
  );
```

but it doesn't seem to work as expected. Inspiration: https://stackoverflow.com/a/53199673/5069352

Is this something already possible or not supported at the moment?

**Comment by enisdenjo** [comment, marked as answer]
When using `useServer`, the [context holds the WebSocket and the initial request in the `extra` prop](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/use_ws.Extra.md).

Sadly, I've never used Passport before, so I cannot tell you what exactly the issue is - but since you're using Cookies, I guess Passport has to be a middleware that is run **before** ever establishing a WebSocket connection.

**Reply by Morishiri** [reply]
The cookies are visible in the request. Seems correct, middleware is attached for regular requests using
```
app.use(passportInitialize);
app.use(passportSession)
```
but it looks like it is never executed in context of websocket connection.

That's why I tried adding calls to it manually in `onConnect` call as proposed re the old approach with `subscription-transport-ws` in linked stackoverflow answer.

Question to clarify, is `onConnect` called before or after upgrade request is done? I mean, if I modify it, is it being used in upgrade request or is it already too late?

**Reply by enisdenjo** [reply]
`onConnect` is called _after_ the WS connection has been established, [read its documentation for more info](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#onconnect).

Can you share your complete server implementation?

**Reply by Morishiri** [reply]
I will try to extract what I cannot share and create some reproductible repository.

**Reply by enisdenjo** [reply]
A repro would be perfect! Thanks.

**Reply by Morishiri** [reply]
Turns out I managed to fix it. I had much more middleware from the `baseAppFactory` we use.

The code I posted in my original question works, the issue in my case was that I had `expressSession` call as well so I needed to modify the code to the following:
```javascript
const sessionInstance = expressSession({
  // options
});
app.use(sessionInstance);

const passportInitialize = passport.initialize();
const passportSession = passport.session();

app.use(passportInitialize);
app.use(passportSession);


const serverCleanup = useServer(
  {
    schema,
    context: async (ctx) => {
       return {
         user: ctx.extra.request.user
       };
    },
    onConnect: async (ctx) =>
      new Promise((resolve) => {
        // Is stored sessionMiddleware() from above
        sessionInstance(ctx.extra.request, {}, () => {
          // Is stored passport.initialize() from above
          passportInitialize(ctx.extra.request, {}, () => {
            // Is stored passport.session() from above
            passportSession(ctx.extra.request, {}, () => {
              resolve(true); // always allow connecting (or check if user is in request already and deny if not, depends what you want to achieve)
            });
          });
        });
      }),
    },
    wsServer,
  );
```

And it was actually mentioned in this stackoverflow answer I posted. May be worth documenting it for the next people having similar issues.

Thanks for your interest and time! :)

**Reply by enisdenjo** [reply]
Glad you got it working! I won't ​add this to the documentation, it is too specific to express and passport - not to `graphql-ws`.

Anyone facing similar issues will either find this Discussion or the answer in StackOverflow.
